src/worker.cpp

```diff
@@ -1185,4 +1183,305 @@ void Worker::collectFrusFromJson()
+        if (!(l_secVpdPtr = std::get_if<types::IPZVpdMap>(&l_parsedSecVpdMap)))
+        {
+            if (std::holds_alternative<std::monostate>(l_parsedSecVpdMap))
+                l_secVpdPtr = new types::IPZVpdMap();
+            else
+            {
+                logging::logMessage("Secondary VPD is not of IPZ type.");
+                return;
+            }
+        }
+
+        backupAndRestoreIpzVpd(*l_parsedVpdPtr, *l_backupMapPtr,
+                               *l_primaryVpdPtr, *l_secVpdPtr);
+    }
+    else if (std::holds_alternative<types::KeywordVpdMap>(
+                 l_parsedPrimaryVpdMap) ||
+             std::holds_alternative<types::KeywordVpdMap>(l_parsedSecVpdMap))
+    {
+        // To Do, handle implementation for Keyword type VPD
+    }
+}
+
+void Worker::backupAndRestoreIpzVpd(types::IPZVpdMap& io_parsedVpdMap,
+                                    const types::IpzBackupMap& i_backupInfoMap,
+                                    const types::IPZVpdMap& i_primaryVpdMap,
+                                    const types::IPZVpdMap& i_secondaryVpdMap)
+{
+    bool l_isParsedMapEmpty = io_parsedVpdMap.empty();
+    const std::string& l_primaryHwPath = getBackupHwPath("primary");
+    const std::string& l_secondaryHwPath = getBackupHwPath("secondary");
+    const auto& l_invPathForPrimary = getInvPathForBackupPath("primary");
+    const auto& l_invPathForSecondary = getInvPathForBackupPath("secondary");
+
+    for (const auto& l_recordKwInfo : i_backupInfoMap)
+    {
+        const std::string& l_recordName = l_recordKwInfo.first;
+        if (!l_isParsedMapEmpty &&
+            io_parsedVpdMap.find(l_recordName) == io_parsedVpdMap.end())
+        {
+            logging::logMessage("Record: " + l_recordName +
+                                ", is not found in the parsed VPD map.");
+            continue;
+        }
+
+        if (!i_primaryVpdMap.empty() &&
+            i_primaryVpdMap.find(l_recordName) == i_primaryVpdMap.end())
+        {
+            logging::logMessage(
+                "Record: " + l_recordName +
+                ", is not found in the primary path: " + l_primaryHwPath);
+            continue;
+        }
+
+        if (!i_secondaryVpdMap.empty() &&
+            i_secondaryVpdMap.find(l_recordName) == i_secondaryVpdMap.end())
+        {
+            logging::logMessage(
+                "Record: " + l_recordName +
+                ", is not found in the secondary path: " + l_secondaryHwPath);
+            continue;
+        }
+
+        const auto& l_kwInfoList = l_recordKwInfo.second;
+        for (const auto& l_keywordInfo : l_kwInfoList)
+        {
+            const std::string& l_keyword = get<0>(l_keywordInfo);
+            const auto& l_defaultBinaryValue = get<1>(l_keywordInfo);
+            bool l_isPelRequired = get<2>(l_keywordInfo);
+
+            const auto& l_primaryBinaryValue = getIpzKeywordValue(
+                i_primaryVpdMap, l_invPathForPrimary, l_recordName, l_keyword);
```

Comment:
If "i_primaryVpdMap" is not empty then by the time execution reaches here, it has already checked if record exist or not. There is no point checking the same inside "getIpzKeywordValue" again.
IMO dont need this API at all, a check if the the map is not empty, directly call "utils::getKwVal" else read from Dbus?
Does this make sense?

Reply:
As earlier comment, I moved getting keyword value into to an API, so that it can be used for getting both primary and secondary value.
record check is being add as generic check in the API, so that it wont be breaking if record not found in the vpd map. so any one can use this API in the future.